Blog post about quantum key exchange (#457)
n0-computer/iroh.computer: src/app/blog/iroh-post-quantum-handshakes/page.mdx

So let's take a look where post-quantum algorithms are relevant for iroh:

During iroh connection establishment, we verify the identity of the remote endpoint. This requires a [*signature algorithm*][wiki-sig]. We exchange a secret for the symmetric encryption of the connection. This requires a [*key exchange algorithm*][wiki-kex].

So there are two attack vectors. If the *signature algorithm* is broken, an attacker can *impersonate* any endpoint id. If the *key exchange algorithm* is broken, an attacker can *decrypt* the contents of the connection.

You might think this isn't a problem. After all, nobody has a sufficiently powerful quantum computer today.

But an attacker can *record* connections now and then break the *key exchange* later, once a sufficiently capable quantum computer becomes available. This is called [*harvest now, decrypt later*][hndl] or *HNDL*.

It's possible that an attacker *listens* to your network (a "passive network attacker") and stores encrypted packets you're sending and receiving in the hopes of being able to decrypt them later, once a cryptographically-relevant quantum computer (CRQC) is built/becomes accessible to the attacker.

> Once that happens, the attacker would go through the key exchange messages using algorithms using RSA or elliptic curves such as P256 or X25519, break their secret keys and be able to derive the symmetric encryption keys used in transmission and thus be albe to decrypt later communication.

Reviewer: ```suggestion
Once that happens, the attacker would go through the key exchange messages using algorithms using RSA or elliptic curves such as P256 or X25519, break their secret keys and be able to derive the symmetric encryption keys used in transmission and thus be able to decrypt later communication.
```
typo

We have to configure iroh to use the `aws-lc-rs` Rust binding to [AWS-LC][aws-lc] via the iroh feature flag `tls-aws-lc-rs`, and configure rustls to *prefer* post-quantum handshakes via the `prefer-post-quantum` rustls feature.

<Note>
The other popular crypto provider [`ring`][ring] as of now does not support post-quantum handshakes.
</Note>

In the `Cargo.toml` of your project, add iroh and rustls with the following feature flags:

```toml
iroh = { version = "0.98", default-features = false, features = ["tls-aws-lc-rs"] }
# iroh depends on rustls with default-features = false, so we need to enable this explicitly
rustls = { version = "0.23.38", default-features = false, features = ["prefer-post-quantum"] }
```

<Note>
What `prefer-post-quantum` actually turns on is [`X25519MLKEM768`][tls-ecdhe-mlkem], a *hybrid* key exchange: both sides run X25519 and [ML-KEM-768][fips-203] and combine the secrets. The session key is safe as long as *at least one* of the two primitives holds — so if ML-KEM turns out to be broken, X25519 still protects you.

> This is necessary because post quantum key exchange algorithms are novel and not as well understood as classical algorithms.

Reviewer: ```suggestion
This is wise to protect from [post quantum key exchange algorithms being broken in classical ways](https://www.quantamagazine.org/post-quantum-cryptography-scheme-is-cracked-on-a-laptop-20220824/), given they are a more recent innovation compared to classic key exchange algorithms.
```

</Note>

This configuration *prefers* a post-quantum handshake despite the higher cost, if *both sides* support it. It does not *force* a post-quantum handshake, so if you talk to an iroh node that does not support post-quantum, you will get a normal X25519 handshake.

So if you are extremely paranoid you would have to configure your iroh endpoint to not allow classical handshakes *at all*, and lose interoperability with default configured iroh endpoints and relays.

Keep in mind that turning on post-quantum key exchange isn't free. `X25519MLKEM768` adds roughly 1 KB in each direction — enough that the handshake requires multiple UDP packets in both directions. `aws-lc-rs` is also a heavier dependency than `ring`.

# What about signatures?

While post-quantum key exchange algorithms are standardized and come with relatively modest overhead, post-quantum signature algorithms are [the subject of active research][nist-pqc-sig].

All current options have [significant downsides][pq-sig-comparison] regarding computational cost, key size or signature size. So there is no industry consensus yet which one to use.

> There is also less urgency because there is no harvest-now-decrypt-later equivalent for signatures.
> 

Reviewer: I'd like to add a small comment about HTTPS certificates also still using classical signatures.

```suggestion
There is also less urgency because there is no harvest-now-decrypt-later equivalent for signatures.

This is similar to the HTTPS world: [More than half of all cloudflare HTTPS connections are now now served using hybrid key exchange algorithms](https://blog.cloudflare.com/pq-2025/), but TLS certificates still entirely depend on classical signature algorithms at the root.

```

Reviewer: Also this user posted some cool stats: https://community.letsencrypt.org/t/lets-encrypt-stats-breakdown-by-signature-algorithm/245420/7

Apparently ~40% of certificates issued still use RSA-1024. 🙃

# Conclusion

Iroh today lets you *opt in* to a post-quantum *key exchange algorithm*, protecting against *harvest now, decrypt later*. If your traffic needs to stay confidential for a few years or more, turn it on.

> Eenabling a quantum-secure key exchange algorithm *does not* protect from active network attackers (a "man in the middle") that intercepts and changes network packets *with access to a CRQC*. Such attacks are impossible at the moment, but might become possible in the future.

Reviewer: ```suggestion
Enabling a quantum-secure key exchange algorithm *does not* protect from active network attackers (a "man in the middle") that intercepts and changes network packets *with access to a CRQC*. Such attacks are impossible at the moment, but might become possible in the future.
```
